R/sgdnet.R

```diff
@@ -0,0 +1,103 @@
+#' Fit a Generalized Linear Model with Elasticnet Regularization
+#'
+#' @param x input matrix
+#' @param y response variable
+#' @param family reponse type
+#' @param alpha elasticnet mixing parameter
+#' @param lambda regularization strength
+#' @param intercept whether to fit an intercept or not
+#' @param maxit maximum number of effective passes (epochs)
+#' @param standardize whether to standardize `x` or not -- ignored when
+#'   `intercept == TRUE`.
+#' @param thresh tolerance level for termination of the algorithm
+#' @param return_loss whether to compute and return the loss at each outer
+#'   iteration, only added here for debugging purposes.
+#' @param ... ignored
+#'
+#' @return An object of class `'sgdnet'`.
+#' @export
+#'
+#' @examples
+#' # Gaussian regression with sparse features
+#' x <- Matrix::rsparsematrix(100, 10, density = 0.2)
+#' y <- rnorm(100)
+#' sgdnet(x, y)
+sgdnet <- function(x, y, ...) UseMethod("sgdnet")
+
+#' @export
+#' @rdname sgdnet
+sgdnet.default <- function(x,
+                           y,
+                           family = c("gaussian"),
+                           alpha = 1,
+                           lambda = 1/NROW(x),
+                           maxit = 1000,
+                           standardize = TRUE,
+                           intercept = TRUE,
+                           thresh = 0.001,
+                           return_loss = FALSE,
+                           ...) {
+
+  n_samples <- NROW(x)
+
+  # The internal optimizer uses a different penalty construction than
+  # glmnet. Convert lambda vallues to match alpha and beta from scikit-learn.
+  alpha_sklearn <- lambda/2*n_samples*(1 - alpha)
+  beta_sklearn <- lambda/2*n_samples*alpha
+
+  # Convert sparse x to dgCMatrix class from package Matrix.
+  if (is_sparse <- inherits(x, "sparseMatrix")) {
+    x <- methods::as(x, "CsparseMatrix")
+    x <- methods::as(x, "dgCMatrix")
```

Comment:
Why the double conversion? Does `as(x, "dgCMatrix")` not automatically work for all `x` classes?

Reply:
I borrowed this from glmnet

https://github.com/cran/glmnet/blob/ff79c8962562e278abfaafb98641a7aecf140b23/R/glmnet.R#L91

It seemed strange to me but I figured it was there for a reason. I'll look into it.

Reply:
No, there doesn't seem to be any reason for this. I am dropping it.